src/Stratis.Bitcoin.Features.Consensus/StakeValidator.cs

```diff
@@ -419,6 +423,65 @@ private uint256 ToUInt256(BigInteger input)
 
             return new uint256(array, false);
         }
+
+        /// <inheritdoc />
+        public bool CheckPOSSignature(BlockSignature signature, uint256 blockHash, Transaction coinStake)
+        {
+            if (signature.IsEmpty())
+            {
+                this.logger.LogTrace("(-)[EMPTY]:false");
+                return false;
+            }
+
+            TxOut txout = coinStake.Outputs[1];
+
+            if (PayToPubkeyTemplate.Instance.CheckScriptPubKey(txout.ScriptPubKey))
+            {
+                PubKey pubKey = PayToPubkeyTemplate.Instance.ExtractScriptPubKeyParameters(txout.ScriptPubKey);
+                bool res = pubKey.Verify(blockHash, new ECDSASignature(signature.Signature));
+                this.logger.LogTrace("(-)[P2PK]:{0}", res);
+                return res;
+            }
+
+            // Block signing key also can be encoded in the nonspendable output.
+            // This allows to not pollute UTXO set with useless outputs e.g. in case of multisig staking.
+
+            List<Op> ops = txout.ScriptPubKey.ToOps().ToList();
+            if (!ops.Any()) // script.GetOp(pc, opcode, vchPushValue))
+            {
+                this.logger.LogTrace("(-)[NO_OPS]:false");
+                return false;
+            }
+
+            if (ops.ElementAt(0).Code != OpcodeType.OP_RETURN) // OP_RETURN)
+            {
+                this.logger.LogTrace("(-)[NO_OP_RETURN]:false");
+                return false;
+            }
+
+            if (ops.Count != 2)
+            {
+                this.logger.LogTrace("(-)[INVALID_OP_COUNT]:false");
+                return false;
+            }
+
+            byte[] data = ops.ElementAt(1).PushData;
+
+            if (data.Length > MaxPushDataSize)
```

Comment:
It was already defined in the `PosBlockSignatureRule.cs`. I moved that definition to `StakeValidator.cs`.

Reply:
No but there is a global check for the size of op_return

Reply:
We could use `TxNullDataTemplate` to perform the checks (leveraging `MAX_OP_RETURN_RELAY`) but that would change the max size of  `OP_RETURN` data from 40 to 80.  Not sure that is something we would like to do as it would affect consensus.